File: library/src/main/java/ir/vasl/library/utils/RequestBodyProgressive.java
package ir.vasl.library.utils;


import android.support.annotation.NonNull;

import java.io.IOException;

import okhttp3.MediaType;
import okhttp3.RequestBody;
import okio.Buffer;
import okio.BufferedSink;
import okio.ForwardingSink;
import okio.Okio;
import okio.Sink;


public class RequestBodyProgressive extends RequestBody {

    private long bytesWritten = 0;

    private Listener listener;

    private RequestBody delegate;

    private int SEND_WINDOW_SIZE_BYTES = (int) Math.pow(2, 16); // 64 KiB

    public RequestBodyProgressive(RequestBody delegate, Listener listener) {
        this.delegate = delegate;
        this.listener = listener;
    }

    @Override
    public MediaType contentType() {
        return delegate.contentType();
    }

    @Override
    public long contentLength() {
        try {
            return delegate.contentLength();
        } catch (IOException e) {
            e.printStackTrace();
        }
        return -1;
    }

    @Override
    public void writeTo(@NonNull final BufferedSink sink) throws IOException {
        Sink streamingSink = new ForwardingSink(sink) {
            @Override
            public void write(@NonNull Buffer source, long byteCount) throws IOException {
                long written = 0L;
                while (written < byteCount) {
                    long writeLength = Math.min(SEND_WINDOW_SIZE_BYTES, byteCount - written);
                    sink.write(source.readByteArray(writeLength));
                    written      += writeLength;
                    bytesWritten += writeLength;
                    listener.onRequestProgress(bytesWritten, contentLength());
                }
            }
        };
        BufferedSink bufferedSink = Okio.buffer(streamingSink);
        delegate.writeTo(bufferedSink);
        bufferedSink.flush();
    }

    public interface Listener {
        void onRequestProgress(long bytesWritten, long contentLength);
    }
}

File: library/src/main/java/com/github/library/utils/RequestBodyProgressive.kt
package com.github.library.utils


import okhttp3.MediaType
import okhttp3.RequestBody
import okio.Buffer
import okio.BufferedSink
import okio.ForwardingSink
import okio.Okio
import java.io.IOException


class RequestBodyProgressive(private val delegate: RequestBody, private val listener: (bytesWritten: Long, contentLength: Long) -> Unit) : RequestBody() {

    private var bytesWritten: Long = 0

    private val sendWindowsByteSize = Math.pow(2.0, 16.0).toInt() // 64 KiB

    override fun contentType(): MediaType? {
        return delegate.contentType()
    }

    override fun contentLength(): Long {
        try {
            return delegate.contentLength()
        } catch (e: IOException) {
            e.printStackTrace()
        }

        return -1
    }

    @Throws(IOException::class)
    override fun writeTo(sink: BufferedSink) {
        val streamingSink = object : ForwardingSink(sink) {
            @Throws(IOException::class)
            override fun write(source: Buffer, byteCount: Long) {
                var written = 0L
                while (written < byteCount) {
                    val writeLength = Math.min(sendWindowsByteSize.toLong(), byteCount - written)
                    sink.write(source.readByteArray(writeLength))
                    written += writeLength
                    bytesWritten += writeLength
                    listener(bytesWritten, contentLength())
                }
            }
        }
        val bufferedSink = Okio.buffer(streamingSink)
        delegate.writeTo(bufferedSink)
        bufferedSink.flush()
    }
}